### ISSUE TITLE
Segfault in device_map='auto' weight dispatch on RTX 5090 D v2 (Blackwell, sm_120, GB202-240)

### ISSUE THREAD

**Issue by XHXIAIEIN** [huggingface/accelerate#3933]
## Bug Description

`AutoModelForCausalLM.from_pretrained(..., device_map="auto")` causes a **Segmentation Fault (exit code 139)** on the **NVIDIA GeForce RTX 5090 D v2**, a China-market Blackwell variant using the **GB202-240** die with compute capability **sm_120**.

The crash occurs during the **weight materialization/dispatch phase** managed by `accelerate`. Basic PyTorch CUDA operations and direct `safetensors` loading to GPU all work correctly, isolating the issue to `accelerate`'s device dispatch logic.

### Key Finding

| Operation | Result |
|---|---|
| `torch.randn(..., device='cuda') @ tensor` | **OK** |
| `safetensors.torch.load_file(path, device='cuda:0')` | **OK** (41 tensors, 3.58 GB) |
| `from_pretrained(..., device_map='auto')` | **Segfault** (exit code 139) |
| `from_pretrained(..., device_map='auto', load_in_4bit=True)` | **Segfault** |

The crash happens early (layers 0–8), ruling out OOM. This is a China-specific **RTX 5090 D v2** GPU variant (GB202-240 die, 24 GB / 384-bit), NOT the standard RTX 5090 (GB202-300, 32 GB / 512-bit).

## Minimal Reproducible Example

```python
import torch
from transformers import AutoModelForCausalLM

# ✅ WORKS: Direct safetensors load to GPU
from safetensors.torch import load_file
tensors = load_file("model-00001-of-00008.safetensors", device="cuda:0")
print(f"Direct load: {len(tensors)} tensors on GPU")  # 41 tensors, 3.58 GB
del tensors

# ✅ WORKS: Manual CPU load → GPU move
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen3-8B", dtype=torch.bfloat16, low_cpu_mem_usage=True)
model = model.to("cuda")  # Works for models that fit in VRAM

# ❌ SEGFAULT: device_map="auto"
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)
# Crashes at: Loading weights: ~20% | Materializing param=model.layers.X...
```

### Also crashes with:
```python
# 4-bit quantization
from transformers import BitsAndBytesConfig
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    quantization_config=BitsAndBytesConfig(load_in_4bit=True, bnb_4bit_compute_dtype=torch.bfloat16),
)

# Explicit max_memory
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-14B",
    device_map="auto",
    torch_dtype=torch.bfloat16,
    max_memory={0: "20GiB", "cpu": "30GiB"},
)
```

## Expected Behavior

`device_map="auto"` should load the model across GPU and CPU without crashing.

## Actual Behavior

Process killed by Segmentation Fault (exit code 139). No Python traceback — the crash is at the C/CUDA level during weight dispatch.

## Environment

| Component | Version |
|---|---|
| **accelerate** | 1.12.0 |
| **PyTorch** | `2.12.0.dev20260223+cu128` (nightly) |
| **CUDA (PyTorch)** | 12.8 |
| **transformers** | latest |
| **bitsandbytes** | 0.49.2 |
| **Python** | 3.14 |
| **OS** | Windows 11 Pro (10.0.26200) |
| **NVIDIA Driver** | 591.86 |
| **GPU** | NVIDIA GeForce RTX 5090 D v2 (GB202-240, 24 GB GDDR7, 384-bit, sm_120) |

## Hardware Details

| Attribute | RTX 5090 (Standard) | RTX 5090 D v2 (this card) |
|---|---|---|
| Die | GB202-300 | **GB202-240** |
| VRAM | 32 GB / 512-bit | **24 GB / 384-bit** |
| Compute Capability | 12.0 | 12.0 |
| Market | Global | **China-specific** |

## Suspected Root Cause

The crash likely occurs in `accelerate.utils.modeling.set_module_tensor_to_device()` or a related function that materializes meta-tensors during weight dispatch. Since direct `safetensors.torch.load_file(..., device='cuda:0')` works, the issue may be in how `accelerate` performs tensor copy/materialization (perhaps using a different CUDA API or code path that is not yet compatible with sm_120/Blackwell).

## Related Issues

- PyTorch: https://github.com/pytorch/pytorch/issues/175614
- Ollama (llama.cpp MMQ kernel crash on same GPU): https://github.com/ollama/ollama/issues/14374

**Comment by XHXIAIEIN** [comment]
## Update: Confirmed Workaround — Bypass `device_map` Entirely

After further testing, I confirmed a reliable workaround that avoids the segfault by bypassing `accelerate`'s weight dispatch:

```python
import torch
from transformers import AutoModelForCausalLM

# ❌ SEGFAULT: accelerate dispatch path
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen3-8B", device_map="auto", torch_dtype=torch.bfloat16)

# ✅ WORKS: Manual CPU → GPU (bypasses accelerate dispatch)
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen3-8B", dtype=torch.bfloat16, low_cpu_mem_usage=True)
model = model.to("cuda")
```

### End-to-End Verification

Ran a full RAG pipeline (Qdrant vector retrieval → Qwen3-8B generation → self-reflection) successfully:

| Metric | Value |
|---|---|
| Model | Qwen/Qwen3-8B (bf16) |
| VRAM | 15.3 GB / 24 GB |
| Speed | ~8.9 tok/s |
| Result | Correct, high-confidence responses |

### Bug Isolation

This further narrows down the issue:

| Operation | Uses accelerate dispatch? | Result |
|---|---|---|
| `safetensors.torch.load_file(path, device='cuda:0')` | No | ✅ Works |
| `from_pretrained(..., low_cpu_mem_usage=True)` + `.to("cuda")` | No | ✅ Works |
| `from_pretrained(..., device_map="auto")` (bf16) | **Yes** | ❌ Segfault |
| `from_pretrained(..., device_map="auto", load_in_4bit=True)` | **Yes** | ❌ Segfault |

The crash is specifically in `accelerate`'s meta-tensor materialization / `set_module_tensor_to_device()` path. All non-accelerate CUDA operations work correctly on this GPU.

### Limitation

This workaround only works for models that fit entirely in VRAM. For models requiring CPU↔GPU split, `device_map` is still needed — and it still segfaults. This means larger models (e.g., 14B+ bf16) cannot be loaded on this 24 GB card at all, since both `device_map="auto"` and `load_in_4bit` (which also uses `device_map`) crash.

### Environment
- accelerate: 1.12.0
- PyTorch: `2.12.0.dev20260223+cu128` (nightly)
- GPU: NVIDIA GeForce RTX 5090 D v2 (GB202-240, sm_120, 24 GB)
- Driver: 591.86 | OS: Windows 11 Pro

**Comment by XHXIAIEIN** [comment]
## Correction & Definitive Verification

### 1. Version Corrections

| Field | Originally reported | Actual |
|---|---|---|
| Python | 3.14 | **3.14.3** |
| transformers | "latest" | **5.2.0** |

### 2. Definitive OOM Exclusion — Tested with Model That Fits in VRAM

The original report only tested with Qwen3-14B (~28 GB bf16, exceeds 24 GB VRAM). This left open the question: is the segfault caused by `device_map` dispatch or by memory overflow during GPU/CPU split?

I re-tested with **Qwen3-8B (~16 GB bf16)**, which fits entirely in 24 GB VRAM with room to spare:

```
Before load:
  VRAM: 20.2 / 23.9 GB free  (model needs ~16 GB → plenty of room)
  RAM:  30.6 GB available

Result:
  device_map="auto" → Segmentation fault (exit code 139)
  Crashed at: 19% (76/399) — model.layers.6.self_attn.o_proj.weight
```

**This confirms the bug is in `accelerate`'s dispatch path, not an OOM condition.** There was no need for CPU offload (model fits in GPU), yet the process still segfaulted during weight materialization.

For comparison, the same model loads successfully with the manual path:
```python
model = AutoModelForCausalLM.from_pretrained(..., low_cpu_mem_usage=True)  # CPU
model = model.to("cuda")  # Manual move — OK
```

### 3. Clarification on 4-bit Case

The original report listed `load_in_4bit=True` as reproducing the "same" segfault. This may actually be a **separate issue** — `bitsandbytes` 0.49.2 includes sm_120 kernels (when paired with cu128), but those kernels may not be fully validated on Blackwell hardware. The 4-bit segfault could be a `bitsandbytes` kernel bug rather than an `accelerate` dispatch bug. I recommend treating them as separate failure modes for triage purposes.
